src/com/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -883,6 +884,106 @@
     (sql/do-commands
      "DROP INDEX fact_values_string_trgm")))
 
+(defn- lift-fact-paths-into-facts
+  "Pairs paths and values directly in facts, i.e. change facts from (id
+  value) to (id path value)."
+  []
+  (sql/do-commands
+   (if (scf-utils/postgres?)
+     "SET CONSTRAINTS ALL DEFERRED" ;; In case it allows additional optimization?
+     "SELECT 1")
+
+   ;; Build complete facts table as of migration 28.
+   "CREATE TABLE facts_transform
+        (factset_id bigint NOT NULL,
+         fact_path_id bigint NOT NULL,
+         fact_value_id bigint NOT NULL)"
+
+   ;; Needed for the insert if nothing else.
+   "CREATE INDEX fact_paths_path_idx on fact_paths(path)"
+   "CREATE INDEX fact_values_value_hash_idx on fact_values(value_hash)"
+
+   ;; Patch up facts refrences to refer to the min id path/value
+   ;; wherever there's more than one option.
+   "INSERT INTO facts_transform (factset_id, fact_path_id, fact_value_id)
+        SELECT f.factset_id,
+               (SELECT MIN(id) FROM fact_paths fp2 WHERE fp2.path = fp.path),
+               (SELECT MIN(id) FROM fact_values fv2
+                  WHERE fv2.value_hash = fv.value_hash)
+          FROM facts f
+            INNER JOIN fact_values fv on f.fact_value_id = fv.id
+            INNER JOIN fact_paths fp on fv.path_id = fp.id"
+   "DROP TABLE facts"
+
+   "ALTER TABLE facts_transform
+        ADD CONSTRAINT facts_factset_id_fact_path_id_fact_key
+          UNIQUE (factset_id, fact_path_id)"
+   "ALTER TABLE facts_transform
+        ADD CONSTRAINT factset_id_fk
+          FOREIGN KEY (factset_id) REFERENCES factsets(id)
```

Comment:
There are a couple of ways to check for covering indexes automatically, PgAdmin3 has a facility. Also this was provided by @dalen: http://mlawire.blogspot.se/2009/08/postgresql-indexes-on-foreign-keys.html?m=1